File: 2015/ts/i-was-told-there-would-be-no-math/main.ts
type Dimensions = {
    length: number;
    width: number;
    height: number;
};

function readInput(path = './input.txt'): Array<Dimensions> {
    const content = Deno.readTextFileSync(path);
    return content.trim().split('\n').map(line => {
        const [length, width, height] = line.split('x').map(v => Number.parseInt(v));
        return {length, width, height};
    });
}

const paperNeeded = (d: Dimensions) => {
    const sides = [d.length * d.width, d.width * d.height, d.height * d.length];
    return 2*(sides[0] + sides[1] + sides[2]) + Math.min(...sides);
};

const ribbonNeeded = (d: Dimensions) => {
    const sides  = [d.length, d.width, d.height].sort((a, b) => a - b);
    return sides[0]*2 + sides[1]*2 + (d.length * d.width * d.height);
};

export function part1(): number {
    return readInput().map(d => paperNeeded(d)).reduce((a, b) => {return a+b}, 0);
}

export function part2(): number {
    return readInput().map(d => ribbonNeeded(d)).reduce((a, b) => {return a+b}, 0);
}

File: 2015/rs/i-was-told-there-would-be-no-math/src/main.rs
use std::fs;

#[derive(Debug, PartialEq)]
struct Dimensions {
    length: i32,
    width: i32,
    height: i32,
}

impl Dimensions {
    fn from_str(s: &str) -> Self {
        let components: Vec<&str> = s.split('x').collect();
        Dimensions {
            length: components[0].parse().unwrap(),
            width: components[1].parse().unwrap(),
            height: components[2].parse().unwrap(),
        }
    }

    fn new(length: i32, width: i32, height: i32) -> Self {
        Dimensions {
            length,
            width,
            height,
        }
    }

    fn paper_needed(&self) -> i32 {
        let sides: [i32; 3] = [
            self.length * self.width,
            self.width * self.height,
            self.height * self.length,
        ];
        let a: i32 = sides.iter().sum();
        let b: i32 = *sides.iter().min().unwrap();
        a * 2 + b
    }

    fn ribbon_needed(&self) -> i32 {
        let mut sides: [i32; 3] = [self.length, self.width, self.height];
        sides.sort();
        sides[0] * 2 + sides[1] * 2 + (self.length * self.width * self.height)
    }
}

#[test]
fn test_dimension_from_string() {
    assert_eq!(Dimensions::from_str("2x3x4"), Dimensions::new(2, 3, 4));
}

#[test]
fn test_paper_needed() {
    assert_eq!(Dimensions::new(2, 3, 4).paper_needed(), 58);
}

fn read_input() -> Vec<Dimensions> {
    let mut results = Vec::new();
    let contents = fs::read_to_string("input.txt").unwrap();
    for line in contents.lines() {
        results.push(Dimensions::from_str(line));
    }
    results
}

//

fn part1() -> i32 {
    let mut total: i32 = 0;
    for dimension in read_input() {
        total += dimension.paper_needed();
    }
    total
}

fn part2() -> i32 {
    let mut total: i32 = 0;
    for dimension in read_input() {
        total += dimension.ribbon_needed();
    }
    total
}

//

fn main() {
    println!("Part one: {}", part1());
    println!("Part two: {}", part2());
}
